core/src/main/java/com/datastax/oss/driver/internal/core/metrics/HdrReservoir.java

```diff
@@ -0,0 +1,240 @@
+import java.io.PrintWriter;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.HdrHistogram.Histogram;
+import org.HdrHistogram.HistogramIterationValue;
+import org.HdrHistogram.Recorder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A reservoir implementation backed by the HdrHistogram library.
+ *
+ * <p>It uses a {@link Recorder} to capture snapshots at a configurable interval: calls to {@link
+ * #update(long)} are recorded in a "live" histogram, while {@link #getSnapshot()} is based on a
+ * "cached", read-only histogram. Each time the cached histogram becomes older than the interval,
+ * the two histograms are switched (therefore statistics won't be available during the first
+ * interval after initialization, since we don't have a cached histogram yet).
+ *
+ * @see <a href="http://hdrhistogram.github.io/HdrHistogram/">HdrHistogram</a>
+ */
+public class HdrReservoir implements Reservoir {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HdrReservoir.class);
+
+  private final String logPrefix;
+  private final Recorder recorder;
+  private final long refreshIntervalNanos;
+
+  // The lock only orchestrates `getSnapshot()` calls; `update()` is fed directly to the recorder,
+  // which is lock-free. Read operations are comparatively rare, so locking is not a bottleneck.
+  private final ReadWriteLock cacheLock = new ReentrantReadWriteLock();
+  private Histogram cachedHistogram; // Guarded by cacheLock
+  private long cachedHistogramTimestampNanos; // Guarded by cacheLock
+  private Snapshot cachedSnapshot; // Guarded by cacheLock
+
+  public HdrReservoir(
+      Duration highestTrackableLatency,
+      int numberOfSignificantValueDigits,
+      Duration refreshInterval,
+      String logPrefix) {
+    this.logPrefix = logPrefix;
+    this.recorder = new Recorder(highestTrackableLatency.toNanos(), numberOfSignificantValueDigits);
+    this.refreshIntervalNanos = refreshInterval.toNanos();
+    this.cachedHistogramTimestampNanos = System.nanoTime();
+    this.cachedSnapshot = EMPTY_SNAPSHOT;
+  }
+
+  @Override
+  public void update(long value) {
+    try {
+      recorder.recordValue(value);
+    } catch (ArrayIndexOutOfBoundsException e) {
+      LOG.warn("[{}] Recorded value ({}) is out of bounds, discarding", logPrefix, value);
+    }
+  }
+
+  /**
+   * <em>Not implemented</em>.
+   *
+   * <p>This reservoir implementation is intended for use with a {@link
+   * com.codahale.metrics.Histogram}, which doesn't use this method.
+   */
+  @Override
+  public int size() {
+    throw new UnsupportedOperationException("HdrReservoir does not implement size()");
+  }
+
+  @Override
+  public Snapshot getSnapshot() {
+    long now = System.nanoTime();
+
+    cacheLock.readLock().lock();
+    try {
+      if (now < cachedHistogramTimestampNanos + refreshIntervalNanos) {
```

Comment:
This can overflow so I think the recommended approach is to use `(now - cachedHistogramTimestampNanos) > refreshIntervalNanos`